crates/ink/tests/ui/contract/fail/constructor-return-result-invalid.stderr

```diff
@@ -0,0 +1,14 @@
+error[E0277]: the trait bound `codegen::dispatch::execution::private::Seal<Result<u8, contract::Error>>: codegen::dispatch::execution::ConstructorReturnType<Contract>` is not satisfied
+  --> tests/ui/contract/fail/constructor-return-result-invalid.rs:14:9
+   |
+14 |         pub fn constructor() -> Result<u8, Error> {
+   |         ^^^ the trait `codegen::dispatch::execution::ConstructorReturnType<Contract>` is not implemented for `codegen::dispatch::execution::private::Seal<Result<u8, contract::Error>>`
+   |
+   = help: the following other types implement trait `codegen::dispatch::execution::ConstructorReturnType<C>`:
+             codegen::dispatch::execution::private::Seal<C>
+             codegen::dispatch::execution::private::Seal<Result<C, E>>
+note: required by a bound in `execute_constructor`
+  --> src/codegen/dispatch/execution.rs
+   |
+   |     private::Seal<R>: ConstructorReturnType<Contract>,
+   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `execute_constructor`
```

Comment:
Ideally those error messages are more user friendly and readable. For example `ConstructorReturnType` could be renamed to `ConstructorReturnsSelfOrResultOfSelf` or something like that.
Also `codegen::dispatch::execution::private::Seal<C>` could rename `C` to `Contract` or even try renaming it to `Self`? Same applies for the line below. I am not sure these are improvements but I am sure we can do better than the current state of error message reports.

Reply:
I personally see no benefit of putting longer name of a trait. The error message is still confusing and unclear.

My proposal is to add a lint for constructor. It will be a follow-up issue

Reply:
I am okay with a proper lint, too, if it significantly improves error messages for smart contracts. I'd hope though that lints are kind of non-optional for ink! since what is the point then?

Reply:
Lints are options with https://github.com/paritytech/cargo-contract/pull/799

Reply:
We have made them optional for now since there currently are none, to avoid having to install `cargo dylint`. This can be switched back to make them compulsory again once we have lints again.

Reply:
Okay understood. Originally I had in mind that those ink! specific lints are mandator and viewed similar to Rust and its borrow checker. This allows to implement additional safety guards into ink! and also rely on them across ink! smart contract dependencies in the same way that Rust crates can depend on the safety guarantees that safe Rust brings with it.

Reply:
Absolutely. Once we have these checks then we can make the linting mandatory again.